_posts/2023-01-22-auto-restarting-connectors.md

```diff
@@ -0,0 +1,328 @@
+---
+layout: post
+title:  "Auto-restarting connectors in Apache Kafka Connect and Mirror Maker 2"
+date: 2023-01-22
+author: jakub_scholz
+---
+
+Apache Kafka Connect provides a framework to integrate Apache Kafka with external systems using connectors.
+Connectors run inside the Connect deployment, connect to the external system and help to push data from Kafka to the external system or the other way around.
+When running Apache Kafka Connect, you have to monitor not only the state of the Connect deployment but also the state of the connectors and their tasks.
+When the connector or one of its tasks fail, you have to analyze the issue to see how to fix it.
+In many cases, you will find out that the issue was caused just by some simple temporary issue such as a network glitch or a temporary outage of the external system.
+And all you need to do to fix it is just restart the connector or task.
+Could this be something the operator might do for you?
+
+<!--more-->
+
+Having the operator automatically restart failed connectors or their tasks was one of the most common feature requests.
+And in Strimzi 0.33, this feature was contributed by [Thomas Dangleterre](https://github.com/ThomasDangleterre).
+So, how does it work?
+
+The auto-restart feature is supported only when using the _connector operator_ and the `KafkaConnector` custom resources to manage the connectors.
+It is disabled by default.
+You can enable it in the `.spec.autoRestart` section of the `KafkaConnector` resource:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnector
+metadata:
+  name: my-connector
+spec:
+  # ...
+  autoRestart:
+    enabled: true
+  # ...
+```
+
+Once enabled, the Strimzi cluster operator will in every reconciliation watch the connector and its tasks for failures.
+And if it sees that any of them are failed, it will automatically restart them.
+
+Not every error can be solved by a restart.
+If the connector or its task failed by some temporary problem such as the network issue mentioned earlier, a restart is an obvious solution.
+But if it is failing because it does not understand the protocol of the remote system or has the wrong credentials, then a restart will not help.
+In these cases, we do not want to keep restarting the connectors forever.
+
+That is why the operator will always restart the connector or its task only up to 7 times.
+The restarts will happen always after an increasing time period.
+The first restart will happen immediately.
+If it does not help, the next restart will happen only after another 2 minutes.
+If even the second restart doesn't help, the next restart will be done only after another 6 minutes.
+And so on.
```

Comment:
I think it may be worth mentioning here when the last restart is expected. I think it should be after 112 minutes since the first restart.

Reply:
Added. Good idea.